File: inst/python/cell2location.py
import scanpy as sc
import anndata
import pandas as pd
import cell2location
import numpy as np
import matplotlib.pyplot as plt
import matplotlib as mpl

#from matplotlib import rcParams
#rcParams["pdf.fonttype"] = 42


# sp_obj = sc.datasets.visium_sge(sample_id="V1_Human_Lymph_Node")
# sp_obj.obs['sample'] = list(sp_obj.uns['spatial'].keys())[0] # add sample information
# # 
# sp_obj.var['SYMBOL'] = sp_obj.var_names # add rownames as column, in this case
# sp_obj.var.set_index('gene_ids', drop=True, inplace=True) # ensembl as rownames???
# 
# adata_ref = sc.read(
#     f'./data/sc.h5ad',
#    backup_url='https://cell2location.cog.sanger.ac.uk/paper/integrated_lymphoid_organ_scrna/RegressionNBV4Torch_57covariates_73260cells_10237genes/sc.h5ad'
# )
# 
# adata_ref.var['SYMBOL'] = adata_ref.var.index
# rename 'GeneID-2' as necessary for your data
# adata_ref.var.set_index('GeneID-2', drop=True, inplace=True)
# 
# delete unnecessary raw slot (to be removed in a future version of the tutorial)
# del adata_ref.raw

def py_build_model_cell2location(adata_ref, 
                                  epochs = 20, 
                                  sample = "sample_id", 
                                  cell_type_column = "celltype_major", 
                                  cell_count_cutoff=5, 
                                  cell_percentage_cutoff=0.03, 
                                  nonz_mean_cutoff=1.12, gpu = False):
                                  
  """
  Build a model using cell2location
  
  Parameters
  ----------
  sc_object: anndata
    AnnData Object containing single-cell reference gene expression 
  sp_object: anndata
    AnnData Object containing spatial expression data
  epochs: integer
    Number of epochs to train the model
  cell_count_cutoff: integer
    Cell2location parameter
  cell_percentage_cutoff: float
    Cell2location parameter
  nonz_mean_cutoff: float
    cell2location parameter
  """

  # filter genes
  # ## mpl.pyplot.ioff()
  selected = cell2location.utils.filtering.filter_genes(adata_ref, 
    cell_count_cutoff = cell_count_cutoff, 
    cell_percentage_cutoff2 = cell_percentage_cutoff, 
    nonz_mean_cutoff = nonz_mean_cutoff)
  adata_ref = adata_ref[:, selected].copy() # filter

  cell2location.models.RegressionModel.setup_anndata(adata=adata_ref,
                          # 10X reaction / sample / batch
                          batch_key=sample,
                          # cell type, covariate used for constructing signatures
                          labels_key=cell_type_column
                         )
                        
  # setup model
  mod = cell2location.models.RegressionModel(adata_ref)
  print ("Training Model")
  mod.train(max_epochs = epochs)
  
  # export model to anndata object
  print ("finished training, extracting results")
  adata_ref = mod.export_posterior(
    adata_ref, sample_kwargs={'num_samples': 1000, 'batch_size': 2500, 'use_gpu': gpu}
  )
  
  # export estimated expression in each cluster
  if 'means_per_cluster_mu_fg' in adata_ref.varm.keys():
    signature = adata_ref.varm['means_per_cluster_mu_fg'][[f'means_per_cluster_mu_fg_{i}'
                                    for i in adata_ref.uns['mod']['factor_names']]].copy()
  else:
    signature = adata_ref.var[[f'means_per_cluster_mu_fg_{i}'
                                    for i in adata_ref.uns['mod']['factor_names']]].copy()
                                    
  # rename celltypes
  signature.columns = adata_ref.uns['mod']['factor_names']

  return signature

#################
# DECONVOLUTION #
#################
def py_deconvolute_cell2location(sp_obj, 
                                 signature, 
                                 epochs= 30000, 
                                 sample = "sample_id", 
                                 n_cell=10, 
                                 alpha=20, 
                                 gpu = True, 
                                 returnValue = "q05"):
  """
  Deconvolute using cell2location 
  
  Parameters
  ----------
  sp_obj: AnnData
    AnnData containing spatial gene expression 
  signature: DataFrame
    DataFrame with gene expression signatures
  epochs: integer
    number of epochs used for trainging
  n_cell: Integer
    Number of expected cells per spot
  alpha: float
    detection alpha
  gpu: boolean
    use gpu acceleration
  """

  # spatial mapping 
  # find shared genes and subset both anndata and reference signatures
  intersect = np.intersect1d(sp_obj.var_names, signature.index)
  sp_obj = sp_obj[:, intersect].copy()
  signature = signature.loc[intersect, :].copy()
  
  # prepare anndata for cell2location model
  cell2location.models.Cell2location.setup_anndata(adata=sp_obj, batch_key=sample)
  
  # create and train the model
  mod = cell2location.models.Cell2location(
      sp_obj, cell_state_df=signature,
      # the expected average cell abundance: tissue-dependent
      # hyper-prior which can be estimated from paired histology:
      N_cells_per_location=n_cell,
      # hyperparameter controlling normalisation of
      # within-experiment variation in RNA detection:
      detection_alpha=alpha
  )
  
  # train model 
  mod.train(max_epochs=epochs,
            # train using full data (batch_size=None)
            batch_size=None,
            # use all data points in training because
            # we need to estimate cell abundance at all locations
            train_size=1,
            use_gpu=gpu)
            
  # export fractions
  sp_obj = mod.export_posterior(
    sp_obj, sample_kwargs = {'num_samples': 1000, 'batch_size': mod.adata.n_obs, 'use_gpu': gpu}
  )
  # means_cell_abundance_w_sf, stds_cell_abundance_w_sf, q05_cell_abundance_w_sf, q95_cell_abundance_w_sf
  
  sp_obj.obs[sp_obj.uns['mod']['factor_names']] = sp_obj.obsm['q05_cell_abundance_w_sf']
  
  return sp_obj.obs[sp_obj.uns["mod"]["factor_names"]]

File: R/Cell2location.R
#' Build a cell2location Model
#'
#' Wraps the cell2location Python workflow via reticulate to train a reference
#' model from single-cell data. Returns the trained model object.
#'
#' @param single_cell_obj `SingleCellExperiment` containing reference expression.
#' @param epochs Number of training epochs.
#' @param assay_sc Single-cell assay to use.
#' @param batch_id_col Batch ID column name.
#' @param cell_type_col Column with cell type labels.
#' @param cell_count_cutoff cell2location parameter.
#' @param cell_percentage_cutoff cell2location parameter.
#' @param nonz_mean_cutoff cell2location parameter.
#' @param gpu Logical; whether to train on GPU.
build_model_cell2location <- function(single_cell_obj, epochs = 250, assay_sc = "counts", batch_id_col = "sample_id", cell_type_col = "celltype_major", cell_count_cutoff = 5, cell_percentage_cutoff = 0.03, nonz_mean_cutoff = 1.12, gpu = TRUE) {
  # build anndata, gene names as rownames

  # init_python()

  ad <- spe_to_ad(single_cell_obj, assay = assay_sc) # using the spatial function

  # source python script
  reticulate::source_python(system.file("python", "cell2location_spacedeconv.py", package = "spacedeconv"))

  model <- py_build_model_cell2location(ad,
    epochs = as.integer(epochs), # int!
    batch_id_col = batch_id_col,
    cell_type_column = cell_type_col,
    cell_count_cutoff = as.integer(cell_count_cutoff), # int!
    cell_percentage_cutoff = cell_percentage_cutoff,
    nonz_mean_cutoff = nonz_mean_cutoff,
    gpu = gpu
  )
}

#' Deconvolute with cell2location
#'
#' Runs the cell2location model (via Python/reticulate) and returns a
#' deconvolution matrix, optionally rescaled to relative values.
#'
#' @param spatial_obj `SpatialExperiment`.
#' @param signature Signature/model input expected by cell2location.
#' @param epochs Training epochs for the model.
#' @param n_cell cell2location hyperparameter.
#' @param alpha cell2location hyperparameter.
#' @param gpu Logical; whether to use GPU for training.
#' @param result_name Prefix used to label result columns (default: "cell2location").
#' @param values `"relative"` to rescale to fractions, or `"absolute"` to keep
#' raw outputs.
deconvolute_cell2location <- function(spatial_obj, signature = NULL, epochs = 30000, n_cell = 10, alpha = 20, gpu = TRUE, result_name = "cell2location", values = "relative") {
  # init_python()

  # TURN INTO ANNDATA
  ad <- spe_to_ad(spatial_obj)

  # source python script
  reticulate::source_python(system.file("python", "cell2location_spacedeconv.py", package = "spacedeconv")) # ("~/spacedeconv/inst/python/cell2location.py")

  deconv <- py_deconvolute_cell2location(
    sp_obj = ad,
    signature = signature, # must be pandas
    epochs = as.integer(epochs),
    n_cell = as.integer(n_cell), alpha = as.integer(alpha), gpu = gpu
  )

  deconv <- attachToken(deconv, result_name)

  if (values == "relative") {
    # abundance_per_spot = rowSums(data.frame(colData(deconv)[, available_results(deconv, method="c2l")]))
    abundance_per_spot <- rowSums(deconv)

    cli::cli_progress_step("Rescaling Cell2location results to relative fractions", msg_done = "Rescaled Cell2location results to relative fractions")
    # for (result in available_results(deconv, method = "c2l")){
    #   colData(deconv)[, result] <- colData(deconv)[, result]/abundance_per_spot
    # }

    for (i in 1:ncol(deconv)) {
      deconv[, i] <- deconv[, i] / abundance_per_spot
    }

    cli::cli_progress_done()
  }

  return(deconv)
}
